### DISCUSSION TITLE
{targets} + {vetiver} (and other kinds of continuous deployment)

### DISCUSSION THREAD

**Discussion by wlandau** [ropensci/targets#1001] [Ideas]
@philbowsher suggested that I explore the power of `vetiver` in `targets` pipelines. (@juliasilge, thanks for the package.) The integration seems like another successful instance of pipeline-based continuous deployment, which also comes up for [Shiny apps](https://www.r-bloggers.com/2020/07/using-drake-for-etl-and-building-shiny-app-for-900k-ct-real-estate-sales/) and literate programming documents. The following `_targets.R` file extends the example from https://github.com/wlandau/targets-four-minutes with a `vetiver` model and and a pin:

```r
# _targets.R file
library(targets)
tar_option_set(packages = c("dplyr", "parsnip", "pins", "vetiver", "workflows"))

pin_model <- function(model) {
  local_options(list(pins.quiet = TRUE))
  board <- board_folder(path = "example_board")
  suppressMessages(vetiver_pin_write(board, model))

  # Return the hash so any target downstream of the pin can update in response to changes:
  pin_meta(board, "air_quality")$pin_hash
}

list(
  tar_target(
    name = data,
    command = filter(airquality, !is.na(Ozone))
  ),
  tar_target(
    name = regression,
    command = lm(Ozone ~ Temp, data)
  ),
  tar_target(
    name = model,
    command = vetiver_model(model = regression, model_name = "air_quality")
  ),
  tar_target(
    name = pin,
    command = pin_model(model)
  )
)
```

The dependency graph shows the overall flow of the pipeline.

```r
# R console
tar_mermaid()
```

```mermaid
graph LR
  subgraph legend
    direction LR
    x0a52b03877696646([""Outdated""]):::outdated --- xbf4603d6c2c2ad6b([""Stem""]):::none
    xbf4603d6c2c2ad6b([""Stem""]):::none --- xf0bce276fe2b9d3e>""Function""]:::none
  end
  subgraph Graph
    direction LR
    xb7119b48552d1da3(["data"]):::outdated --> x2af8164e6b30c875(["regression"]):::outdated
    xe1eeca7af8e0b529(["model"]):::outdated --> xd4981b1ef5777538(["pin"]):::outdated
    x7b4395794557f035>"pin_model"]:::outdated --> xd4981b1ef5777538(["pin"]):::outdated
    x2af8164e6b30c875(["regression"]):::outdated --> xe1eeca7af8e0b529(["model"]):::outdated
  end
  classDef outdated stroke:#000000,color:#000000,fill:#78B7C5;
  classDef none stroke:#000000,color:#000000,fill:#94a4ac;
  linkStyle 0 stroke-width:0px;
  linkStyle 1 stroke-width:0px;
```

And the targets in the pipeline run in topological order.

```r
# R console
tar_make()
#> • start target data
#> • built target data [0.406 seconds]
#> • start target regression
#> • built target regression [0.005 seconds]
#> • start target model
#> • built target model [0.019 seconds]
#> • start target pin
#> • built target pin [0.101 seconds]
#> • end pipeline [0.67 seconds]
```

The pin only updates when the vetiver model object changes. 

```r
# After nothing changed:
tar_make()
#> ✔ skip target data
#> ✔ skip target regression
#> ✔ skip target model
#> ✔ skip target pin
#> ✔ skip pipeline [0.079 seconds]
```

This cuts down on the number of superfluous versions in the model history. In production, [the pipeline could run on Connect](https://solutions.posit.co/gallery/targets/) at regular intervals, maybe in a scenario where the data updates on a schedule and the model needs to be retrained. `targets` would strategically reduce both computation time and the number of unimportant saved models.

Model cards would work well with the literate programming integration at <https://books.ropensci.org/targets/literate-programming.html#literate-programming-within-a-target>. As for `plumber` APIs, another target could redeploy the API with a short-lived synchronous command if needed, and then downstream targets could run predictions with the updated API using `predict(vetiver_endpoint("https://..."), ...)`.

**Comment by juliasilge** [comment]
This is looking great @wlandau! 🙌 Thanks for sharing. Let me know if you have questions as you work on this in the future.
